### PR TITLE
8256450: Add gz option to jmap to write a gzipped heap dump

### PR DESCRIPTION
**Pull request by linzang**
This PR add "gz" option to jmap -dump command to support generate gzipped heap dump.

example:
      jmap -dump:live,gz=1,file=dump.gz <pid>

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (6/6 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |     |     |  ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8256450](https://bugs.openjdk.java.net/browse/JDK-8256450): Add gz option to jmap to write a gzipped heap dump

### Reviewers
 * [Chris Plummer](https://openjdk.java.net/census#cjplummer) (@plummercj - **Reviewer**)
 * [Serguei Spitsyn](https://openjdk.java.net/census#sspitsyn) (@sspitsyn - **Reviewer**)
 * [Paul Hohensee](https://openjdk.java.net/census#phh) (@phohensee - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1251/head:pull/1251`
`$ git checkout pull/1251`

### CODE REVIEW

**Review by plummercj — COMMENTED** [review]
Are the Attach API changes backwards compatible? You've added a new arg, but that arg could be passed to an older JVM that doesn't support it (I think it just gets ignored), or an older JVM would fail to pass the arg to a new JVM that is expecting it.

Also, please see my comments in [JDK-8256451](https://bugs.openjdk.java.net/browse/JDK-8256451) regarding SA heap dumping support, and how it will lack this ability to compress the heap dump.

**Comment by linzang** [comment]
Hi @plummercj, 
     Thanks for your comments! 

> Are the Attach API changes backwards compatible? You've added a new arg, but that arg could be passed to an older JVM that doesn't support it (I think it just gets ignored), or an older JVM would fail to pass the arg to a new JVM that is expecting it.

Correct, the new "jmap -dump:gz=1 [pid]" could work with old JVM and the "gz" option is ignored.  And the old jmap -dump can not accept "gz" option, it fails with error message printed, no matter what jvm version it work with.  I also tested that jcmd GC.heap_dump command with old version of jcmd can not accept the option "-gz=[number]". Moreover, This behavior is similar to "parallel=<num>" option recently added to jmap -histo. So IMHO the risk is not high.

If the "gz" option is not used, jmap -dump could work normally, with different version of jmap command and JVM. (I only tested with jdk8, jdk11 and latest jdk build)

> Also, please see my comments in [JDK-8256451](https://bugs.openjdk.java.net/browse/JDK-8256451) regarding SA heap dumping support, and how it will lack this ability to compress the heap dump.

Yes, I am also considering adding compression support to them, but I am not sure whether there should be new CSR issues created separately to tracking them, and whether I should create those issues at present because I am only planing the work. Let's discuss in the CSR. 
Thanks!

BRs,
Lin

**Comment by plummercj** [comment]
> And the old jmap -dump can not accept "gz" option, it fails with error message printed, no matter what jvm version it work with.

I just want to clarify what I was referring to. I was not talking about trying to use gz with the old jmap command (from the command line). I was asking what happens if you use the old jmap command on a newer jvm that does support (and I assume expects) the new "compression level" argument to be passed, but it won't be. We've had this issue before. See [JDK-8219721](https://bugs.openjdk.java.net/browse/JDK-8219721). I haven't had a chance to look at the fix for JDK-8219721 to see if it also solves the issue with this change, or if something similar also needs to be done with this change.

**Comment by linzang** [comment]
Hi @plummercj ,

Thanks for your explaination. I get your point. 
The issue of JDK-8219721 was fixed by reverting the problematic change. And the root cause of that issue is there is fixed limitation of number of arguments that passing to jvm from attacher, I have figured out a way to avoid touching it when adding new options to jcmd-alike tools, However this PR doesn't even exceed the argument limitation, so it doesn't have chance to cause the similar issue. 

BRs,
Lin

**Comment by plummercj** [comment]
So I assume then that the `op->arg(2)` reference will return `NULL` when and older `jmap` is used? Is that correct?

**Comment by linzang** [comment]
Hi @plummercj,
     For old jmap, if "gz" option is used in commandline, it prompt usage info indicating that option is illegal.
     if "gz" option is not used, the op->arg(2) will be an empty string "", which is set from the attacher side in the method execute() of VirtualMachineImpl.java.

**Comment by plummercj** [comment]
Ok, I see now that 3 args are always written, even if empty, so you at least get a "" if there is no value for the argument position.

**Comment by plummercj** [comment]
I think you need to add a test case. You can use the work done for JDK-8237354 dcmd as a template. See http://cr.openjdk.java.net/~rschmelter/webrevs/8237354/webrev.3/. I don't know that's it's necessary to add a test case for each GC, but if you do, I suggest putting them all in one file. The use of separate files per test case was not necessary.

**Comment by linzang** [comment]
> I think you need to add a test case. You can use the work done for JDK-8237354 dcmd as a template. See http://cr.openjdk.java.net/~rschmelter/webrevs/8237354/webrev.3/. I don't know that's it's necessary to add a test case for each GC, but if you do, I suggest putting them all in one file. The use of separate files per test case was not necessary.

I extended the BasicJMapTest.java to test the compressed heap dump. would you like to help review?
Thanks!

File: test/jdk/sun/tools/jmap/BasicJMapTest.java [hunk 1]
```diff
@@ -237,25 +246,42 @@ private static void dump(boolean live, boolean explicitAll) throws Exception {
...
         output = jmap(allArgs);
         output.shouldHaveExitValue(0);
         output.shouldContain("Heap dump file created");
-        verifyDumpFile(file);
+        verifyDumpFile(file, compressed);
         file.delete();
     }
 
-    private static void verifyDumpFile(File dump) {
+    private static void verifyDumpFile(File dump, boolean compressed) {
```
Comment (plummercj) [inline test/jdk/sun/tools/jmap/BasicJMapTest.java:270]: I don't think the `compressed` argument is needed. The rest of the changes look fine.

**Review by plummercj — CHANGES_REQUESTED** [review]

**Review by plummercj — APPROVED** [review]

**Review by sspitsyn — APPROVED** [review]
Hi Lin,
The fix looks okay to me.
Thanks,
Serguei

**Comment by linzang** [comment]
Thank @plummercj and @sspitsyn  for your help reviewing this PR. I will add /integrate label, and may I ask your help to sponse pushing it? 

BRs,
Lin

**Review by phohensee — APPROVED** [review]

**Pull request closed without merging**